modules/openapi-generator/src/main/resources/python/model_test.mustache

```diff
@@ -5,14 +5,21 @@
 from __future__ import absolute_import
 
 import unittest
+import datetime
```

Comment:
Please delete some of the existing models and have your template generate new examples. How about tests for the below models?
- samples/client/petstore/python/test/test_animal.py (discriminator model)
- samples/client/petstore/python/test/test_cat.py (composed allof model)
- samples/client/petstore/python/test/enum_class.py (model with enums)
- samples/client/petstore/python/petstore_api/models/format_test.py (normal model with many parameter types)

Reply:
I will have no time for that.

Reply:
The request up above will take very little time.
To do it one would run:
```
rm samples/client/petstore/python/test/test_animal.py
rm samples/client/petstore/python/test/test_cat.py
rm samples/client/petstore/python/test/test_enum_class.py
rm samples/client/petstore/python/test/test_format_test.py
bin/python-petstore.sh.sh
# turn on the test server
docker run -d -e SWAGGER_HOST=http://petstore.swagger.io -e SWAGGER_BASE_PATH=/v2 -p 80:8080 swaggerapi/petstore
# test the updated client
cd samples/client/petstore/python
# run the tests
make test
```
That should take around a minute max.

These api_test and model_test updates will impact all of the tests that are auto generated when the users run the generator.
This pull request does not yet include any tests of this new functionality.
Tests will need to be added showing that both the new model tests and the api client tests  produce code which runs and passes tests.
If these tests were omitted this update could breaking the tests produced in the python client. If this is too much work for this PR the model_test update and the api_test update can be broken up in to different PRs.

I am hearing that you are busy. We can wait until you have time to add this test verification. It is very important that new changes do not break the existing tests, so we can go at whatever speed works for you .

Reply:
Another thing that we can do is look at the content of model_test.mustache in other generators and see how they handle instantiating parameters that have model values. Maybe they have some cool tricks that we can use.

Reply:
Hi! I tried to remove and re-create all the existing `samples/client/petstore/python/test/*`, and the hardest tests to manage are exactly the ones with discriminators, since python and python-experimental produce different signatures. For example, for petstore/models/Cat.py:
python:     `def __init__(self, declawed=None, local_vars_configuration=None):`
python-experimental:    `def __init__(self, class_name, _check_type=True, _from_server=False, _path_to_item=(), _configuration=None, **kwargs):`

The funny thing is that class_name (used in the discriminator) is NOT in the old generator, and it is required (it has no default) in the python-experimental generator.

The only solution I see is to develop a model_test.moustache for the old python, and a model_test_experimental.moustache for the python-experimental.

Any suggestions on alternative solutions?

Reply:
Cat is a composed schema model inherriting from Pet and Animal I believe. The Python generator does not fully support composed schema models and the Python-experimental generator does support it. So the Python-experimental model correctly shows that class_name is required. 

If you did want to write a solution that works for both you could iterate over requiredVars and optionalVars which will include class_name. But it will leave out oneOf inherrited schema input variables. This will be painful to implement for composed schemas.

Your proposed solution of using a separate template for python and python-experimental is what we should do. Making examples for composed schema models will be difficult and should probably be implemented in a later PR.
Please break it up into the separate templates.

Reply:
Done. I have completed the code for the python generator, and I will do something later for the python-experimental generator.

Whenever there is a discriminator over a variable of a model (e.g.: allOf), the client generated by the python generator does not have the discriminating property in its #vars, but it has it in #requiredVars. In the generated code, the __init__ of the models cannot accept the discriminating string, but the tests (which are based on the #requiredVars) try to pass the discriminating string, leading to an error in the tests. Moreover, the schema is not of "ComposedSchema" type.
I solved that by looping over #vars. This would make code generation fail if there is a circular dependency in the variables.

Reply:
Can you include the below new test files?
- test_pet.py
- test_file_schema_test_class.py
Please make sure those tests make include the child parameter instances.
So for example in pet we should have:
`category = TestCategory.make_instance(include_optional=include_optional)` NOT `category = None`

Reply:
Are you sure about that? There are two options to do that, both of them having problems. Let us consider a Pet that contains an object variable whose schema is a reference to '#/definitions/Category'.

- from the python code. I would have to hardcode the `TestCategory.make_instance(include_optional=include_optional)` object creation. Drawbacks:
  + Whenever production code asks for an example of Category, it will receive a dependency on the class `TestCategory`.
  + I will have to change the signature of `toExampleValue` to pass the data type around. In fact, current None are caused by references, and with references the second call to toExampleValue, which dereferences the type, does not bring around the model title.

- working on the templates.
I would be able to isolate invocations on `TestCategory` on the tests only, but I will have to process the models into the templates. Drawbacks:
  + I would have to undo most of the work done until now to create examples on-the-fly into the java code. In this moment, the template just asks for an example, leading to a call to toExampleValue.
  + There is still the problem that the model does not contain the internal structure of the objects being instantiated when there is a reference.

Thus, do you suggest to approach the problem using java or the templates?

Reply:
I am hearing that you are too busy to meet.
So the implementation is up to you, but I am sure that I want our Pet model and File Schema Test Class model to contain instances of other models.

Including child model references
- from the python code
   * yes we will have a dependence on TestCategory or Category here. That's okay. The definition of the datatype for category is pointing us to the Category model conceptually.
   * so I don't think that you need to change toExample value to pass in dataType, dataType is already contained in the Schema, or we can get it from converting the schema to a codegenProperty or codegenParameter

- changing the templates
   * I'm not sure exactly what you mean here. It sounds like you would change how examples are put in using the tags when looping over vars.
   * Yes, you don't have the required data here to instantiate required models like Category
   * In my opinion you should leave the templates as-is where they use example values

So I think that there are two solutions here, and we can use the Java toExample method to get to the final python code.
The final python code could look something like:
```
# Option 1
pet = Pet(
  category=petstore_api.Category(
    id=1,
    name='blah',
  ),
  name='blah',
  photo_urls=['blah'],
  tags=[
    petstore_api.Tag(
      id=1,
      name='blah',      
    )
  ],
  status='blah',
)
# Option 2
pet = Pet(
  category=TestCategory.make_instance(include_optional=include_optional),
  name='blah',
  photo_urls=['blah'],
  tags=[TestTag.make_instance(include_optional=include_optional)],
  status='blah',
)
```
In both cases, the above code would be in the make_instance method.
### Option 1
- pro: this can be pasted in by a user into a different application which is great
- con: this requires instantiating codegenModels when we recursively call toExample value which is complicated

### Option 2
- con: if we paste in this code we will also need to cut and paste in the code to replace the TestModelName.make_instance references, so pasting code from one place doesn't work for our uses
- pro: the Java code to do this is easier to write in toExample value because we don't need to instantiate CodegenModels

Updating your toExample function to detect custom models and recursively call itself on array/map parameters is doable.
- Here is where python-experimental recursively defines dataType using an input Schema: https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java#L812
-  Here is where python-experimental detects that a codegenParamter is a model type: https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java#L887

Either Option1 or Option2 is fine, and both can use your existing toExample.. method.
Which do you prefer/what do you think?

Reply:
How should I manage the array of array? For example, if you want to manage this definition correctly:
```yaml
    Bird:
      type: object
      properties:
        photos:
          type: array
          items:
            type: array
            items:
              type: string
```
you should produce an array of arrays of strings, but to do that it is necessary to use recursion. Thus, I need either to do recursion on java side (toExampleValue(Schema schema) calling itself), or play with templates, meaning that I will have a arrayobject.mustache that can do {{>arrayobject}} when needed.

Or I can just decide that, if there is an array of arrays, the internal arrays are [].

Let me know.

Reply:
Please use recursion in the java code. It is a more robust solution and will give our users higher quality examples which include nested values rather than empty list.
You could use this example here: https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java#L812
as a starting point.

Reply:
One more question. If I use Java's toExampleValue, how do I pass it the include_optional parameter? Should I consider it always True, or always False?

Reply:
If you use include_optional=include_optional then you will pass through whatever value was input

Reply:
When I do `{{#example}}{{{.}}}{{/example}}` in the templates, I receive a string from toExampleValue and I use it to build the python file. I create 2 flavors of make_instance, one when optional parameters are included and one not.

Anyway, I don't see any way to send the parameter `include_optional`, which is known at runtime, to toExampleValue that is at code generation.

Reply:
Aren't we writing about examples for individual parameters? The parameter itself is either optional or required which is decided in the mustache template. For individual parameters we then use the example value. The function that uses the example already has include_optional defined when we call it so in the example values we can include the string include_optional=include_optional

Reply:
No, I mean when generating the code. For example:
```yaml
  Tiger:
    type: object
    required:
      - teeth
      - cat
    properties:
      teeth:
        type: string
      cat:
        $ref: '#/definitions/Cat'
  Cat:
    type: object
    required:
      - fur
    properties:
      fur:
        type: string
      animal:
        $ref: '#/definitions/Animal'
  Animal:
    type: object
    required:
      - color
    properties:
      color:
        type: string
```

What do you do, when generating the make_instance for Tiger? The options are to generate:
```
            return Tiger(
                teeth = 'many', 
                cat = openapi_client.models.category.Cat(
                fur = 'short',
                animal = openapi_client.models.category.Animal(
                color = 'black and yellor',
            )))

```
or
```
            return Tiger(
                teeth = 'many', 
                cat = openapi_client.models.category.Cat(
                fur = 'short',
            ))
```
Ideally, you would generate the first when include_optional is True and the second then include_optional is False. Anyway, the code for the Cat is generated by toExampleValue, which does NOT know what include_optional is. Thus, internal properties are either generated ALWAYS with all optional parameters, or NEVER with optional parameters.

And now, here is the bomb:
```yaml
  TagList:
    type: object
    required:
      - name
    properties:
      name:
        type: string
      nextTag:
        $ref: '#/definitions/TagList'
```
It is a linked list, and it should have an empty nextTag sooner or later. Anyway, if we always comprise optional parameters in toExampleValue, we will have infinite recursion.
So, we should NEVER add optional parameters at toExampleValue level. Do you agree?

Reply:
So I see two solutions to these problems.
For the problem of needing to pick whether optional is required or not
- create a string property in CodegenProperty called exampleOptional. In it store the example where include_optional is true. In example store the example where include_optional is false

For the problem of a model with params or items of type model, you can include a list of Schemas or dataTypes that you are inspecting in your toExample function. Every time you call it recursively, add a schema. If we are going to add a schema that we have already added, add it then return, which will end our recursive calls.

Reply:
The first solution implies changing the data model, and I prefer to leave it to the next developer.

I have implement the second solution.

Reply:
Optionally for the first solution we could do it without changing the data model.
This would work by storing the exampleOptional string value in a property that we are not using in CodegenProperty. Why not store that value in the unescapedDescription property and add a comment describing what we are storing in it?

Reply:
Yep, I would prefer to have this PR accepted, then do something for the api, then I could come back to this issue - after discussing it with you in a chat.